pkg/controller/pcideviceclaim/pcideviceclaim_controller.go

```diff
@@ -100,118 +134,204 @@ func unbindPCIDeviceFromVfioPCIDriver(addr string) error {
 	return nil
 }
 
-func (h Handler) reconcilePCIDeviceClaims(nodename string) error {
-	// Get all PCI Device Claims
-	var pdcNames map[string]string = make(map[string]string)
-	pdcs, err := h.pdcClient.List(metav1.ListOptions{})
+func addHostDeviceToKubeVirtAllowList(pd *v1beta1.PCIDevice) error {
+	logrus.Infof("Adding %s to KubeVirt Allow-List", pd.Name)
+	clientConfig := kubecli.DefaultClientConfig(&pflag.FlagSet{})
+	virtClient, err := kubecli.GetKubevirtClientFromClientConfig(clientConfig)
 	if err != nil {
-		return err
+		msg := fmt.Sprintf("cannot obtain KubeVirt client: %v\n", err)
+		return errors.New(msg)
+	}
+	ns := "harvester-system"
+	cr := "kubevirt"
+	kv, err := virtClient.KubeVirt(ns).Get(cr, &v1.GetOptions{})
+	if err != nil {
+		msg := fmt.Sprintf("cannot obtain KubeVirt CR: %v\n", err)
+		return errors.New(msg)
+	}
+	kvCopy := kv.DeepCopy()
+	if kv.Spec.Configuration.PermittedHostDevices == nil {
+		kvCopy.Spec.Configuration.PermittedHostDevices = &kubevirtv1.PermittedHostDevices{
+			PciHostDevices: []kubevirtv1.PciHostDevice{},
+		}
+	}
+	permittedPCIDevices := kv.Spec.Configuration.PermittedHostDevices.PciHostDevices
+	resourceName := pd.Status.ResourceName
+	// check if device is currently permitted
+	var devPermitted bool = false
+	for _, permittedPCIDev := range permittedPCIDevices {
+		if permittedPCIDev.ResourceName == resourceName {
+			devPermitted = true
+		}
+	}
+	vendorId := pd.Status.DeviceId
+	deviceId := pd.Status.DeviceId
+	if !devPermitted {
+		devToPermit := kubevirtv1.PciHostDevice{
+			PCIVendorSelector:        fmt.Sprintf("%s:%s", vendorId, deviceId),
+			ResourceName:             resourceName,
+			ExternalResourceProvider: false,
+		}
+		kvCopy.Spec.Configuration.PermittedHostDevices.PciHostDevices = append(permittedPCIDevices, devToPermit)
+		_, err := virtClient.KubeVirt(ns).Update(kvCopy)
+		if err != nil {
+			msg := fmt.Sprintf("Failed to update kubevirt CR: %s", err)
+			return errors.New(msg)
+		}
 	}
+	return nil
+}
+
+func pciDeviceIsClaimed(pd *v1beta1.PCIDevice, pdcs *v1beta1.PCIDeviceClaimList) bool {
 	for _, pdc := range pdcs.Items {
-		// Build up mapping
-		nodeAddr := fmt.Sprintf(
-			"%s-%s", pdc.Spec.NodeName, pdc.Spec.Address,
-		)
-		pdcNames[nodeAddr] = pdc.Name
+		if pdc.OwnerReferences[0].Name == pd.Name {
+			return true
+		}
+	}
+	return false
+}
+
+// A PCI Device is considered orphaned if it is bound to vfio-pci,
+// but has no PCIDeviceClaim. The assumption is that this controller
+// will manage all PCI passthrough, and consider orphaned devices invalid
+func (h Handler) unbindOrphanedPCIDevices(nodename string) {
+	pdcs, err := h.pdcClient.List(metav1.ListOptions{})
```

Comment:
without return when error, it's not fit to access `pdcs`, `pds` later, as they are nil